components/raftstore/src/store/peer.rs

```diff
@@ -4398,6 +4405,30 @@ where
             );
         }
     }
+
+    pub fn need_renew_lease_at<T>(
+        &self,
+        ctx: &PollContext<EK, ER, T>,
+        renew_bound: Timespec,
+    ) -> bool {
+        let max_lease = ctx.cfg.raft_store_max_leader_lease();
+        match self.leader_lease.inspect(Some(renew_bound)) {
+            LeaseState::Expired => {
+                self.pending_reads.back().map_or(true, |read| {
```

Comment:
Perhaps these can be checked before inspecting lease, so that a query to time can be saved.